[WFLY-14310] Use standard idioms for handling attribute resolution in…

bstansberry opened the pull request:
… JPASubsystemAdd

Remove the 'default-datasource=""' xml attribute from our standard configs, which was only necessary as a workaround of the fact that JPASubsystemAdd was incorrectly calling 'require' to read an optional attribute.

This also does a bit of code cleanup in the class and associated ResourceDefinition.

https://issues.redhat.com/browse/WFLY-14310

bstansberry committed:
[WFLY-14310] Use standard idioms for handling attribute resolution in JPASubsystemAdd

Remove the 'default-datasource=""' xml attribute from our standard configs, which was only necessary as a workaround of the fact that JPASubsystemAdd was incorrectly calling 'require' to read an optional attribute.

Remove invalid parser enforcement of the presence of default-datasource in the xml.

Correct xml marshaller handling of a subsystem with no defined attributes.

This also does a bit of code cleanup in the JPASubystem class, the associated ResourceDefinition and the parsers.

bstansberry commented:
@scottmarlow Please review.

scottmarlow commented:
@bstansberry are the test failures caused by this change?

bstansberry commented:
@scottmarlow I'm going to assume so.

scottmarlow commented:
I'll try building this locally.

scottmarlow commented:
I'm running a few of the JPA tests locally with https://github.com/scottmarlow/wildfly/commit/ad0fbfbaf9babf47c6a2387f809793c1dbb042ae

bstansberry commented:
Thanks, @scottmarlow. Yes, that check should not be there as the xsd records the attribute as optional. And the rest of my PR removes the Stage.RUNTIME code that required it to be there despite what the xsd and attribute definition say.

This will teach me not to run the smoke tests before pushing something up.

I'll add a subsystem test of an empty subsystem element.

scottmarlow commented:
@bstansberry the local JPA tests are passing for me and the management console/jboss-cli output looks good (standalone.xml has expected content after setting default datasource and also correct after removing default datasource).

bstansberry commented:
@scottmarlow I've updated this. Adding new subsystem tests revealed a problem in how the xml is written if the none of the optional attributes are defined; I fixed that as well.

bstansberry commented on jpa/subsystem/src/main/java/org/jboss/as/jpa/subsystem/JPADefinition.java:
```diff
@@ -53,6 +53,23 @@
+                    .setValidator(new EnumValidator<ExtendedPersistenceInheritance>(ExtendedPersistenceInheritance.class,true,true))
+                    .setDefaultValue(new ModelNode(ExtendedPersistenceInheritance.DEEP.toString()))
+                    .build();
+
```
Note for the historical record: The reason this code is moved is because it's all static initialization, and the order in which that is done matters. We want these ADs initialized before the INSTANCE field on L73 is initialized, as that initialization in turn calls code that reads these AD variables.
